### PR TITLE
os/bluestore: rewrite deferred write handling

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-04-13_22:35:11-rados-wip-sage-testing---basic-smithi/

**Comment by liewegas** [comment]
retest this please

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -3165,7 +3267,11 @@ void *BlueStore::MempoolThread::entry()
 static void aio_cb(void *priv, void *priv2)
 {
   BlueStore *store = static_cast<BlueStore*>(priv);
-  store->_txc_aio_finish(priv2);
+  if ((unsigned long long)priv2 & 0x1ull) {
+    store->deferred_aio_finish((void*)((unsigned long long)priv2 & ~1ull));
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: What's about having more universal and less error-prone approach here -> inherit both TransContext & DeferredBatch from an abstract class, implementing single virtual callback method:
struct AioContext{
virtual void  aio_finish(BlueStore*) = 0;
};

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -7983,106 +8104,99 @@ void BlueStore::_deferred_try_submit()
...
-  assert(osr->deferred_running.empty());
+  dout(10) << __func__ << " osr " << osr
+	   << " " << osr->deferred_pending->iomap.size() << " ios pending "
+	   << dendl;
+  assert(osr->deferred_pending);
+  assert(!osr->deferred_running);
 
-  deferred_queue_size -= osr->deferred_pending.size();
+  auto b = osr->deferred_pending;
+  deferred_queue_size -= b->seq_bytes.size();
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: It looks like deferred_queue_size tracks amount of txcs, not bytes. See ++deferred_queue_size in _deferred_queue()

**Comment by liewegas** [comment]
seq_bytes is a map<> that has an entry for each txc

**Comment by liewegas** [comment]
Updated to use AioContext

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -7964,16 +8072,29 @@ bluestore_deferred_op_t *BlueStore::_get_deferred_op(
 
 void BlueStore::_deferred_queue(TransContext *txc)
 {
-  dout(20) << __func__ << " txc " << txc << " on " << txc->osr << dendl;
+  dout(20) << __func__ << " txc " << txc << " osr " << txc->osr << dendl;
   std::lock_guard<std::mutex> l(deferred_lock);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Just some thoughts I want to share - having such a common lock might probably impact the performance negatively - it seems the only global entity really needed such a protection is deferred_queue itself. Other stuff is per-OSR and hence might have another lock with narrower scope. May be refactor later?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -7983,106 +8104,99 @@ void BlueStore::_deferred_try_submit()
...
 {
-  bluestore_deferred_transaction_t& wt = *txc->deferred_txn;
-  dout(20) << __func__ << " txc " << txc << " seq " << wt.seq << dendl;
+  dout(10) << __func__ << " osr " << osr << dendl;
+  assert(osr->deferred_running);
+
+  DeferredBatch *b = osr->deferred_running;
+  {
+    std::lock_guard<std::mutex> l2(osr->qlock);
+    std::lock_guard<std::mutex> l(kv_lock);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Do we really need this lock here?

Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: hmm we could move it down, but we need it for the deferred_done_queue.emplace.

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: I agree, but the locking gets pretty complicated.  I prefer to make that a follow-on change (when we see this as a point of contention)

**Pull request merged**